### PR TITLE
fix(dav): Rate limit address book creation

### PR DESCRIPTION
**Pull request by hamza221**
* Resolves: # 

## TODO

- [x] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+			// Path looks like addressbooks/users/username/addressbooksname so a new addressbook is created
+			try {
+				$this->limiter->registerUserRequest(
+					'carddav-create-address-book',
+					$this->config->getValueInt('dav', 'rateLimitAddressBookCreation', 10),
+					$this->config->getValueInt('dav', 'rateLimitPeriodAddressBookCreation', 3600),
+					$user
+				);
+			} catch (RateLimitExceededException $e) {
+				throw new TooManyRequests('Too many calendars created', 0, $e);
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:90]: ``
```suggestion
				throw new TooManyRequests('Too many addressbooks created', 0, $e);
```

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			}
+
+			$addressBookLimit = $this->config->getValueInt('dav', 'maximumAdressbooks', 10);
+			if ($addressBookLimit === -1) {
+				return;
+			}
+			$numAddressbooks = $this->cardDavBackend->getAddressBooksForUserCount('principals/users/' . $user->getUID());
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:97]: Is this a principal URI? I don't remember :thinking:

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * @copyright 2023 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @author 2023 Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:8]: Hello Christoph

Comment (hamza221) [reply apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:8]: That's on purpose haha

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php (hunk 1 above)
Comment (hamza221) [reply apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:90]: I didn't ctrl C , ctrl v ,  I promise 🤞🏼

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	private IAppConfig $config;
+	private LoggerInterface $logger;
+	private ?string $userId;
+
+	public function __construct(Limiter $limiter,
+		IUserManager $userManager,
+		CardDavBackend $cardDavBackend,
+		LoggerInterface $logger,
+		IAppConfig $config,
+		?string $userId) {
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:55]:
```suggestion
	public function __construct(private Limiter $limiter,
		private IUserManager $userManager,
		private CardDavBackend $cardDavBackend,
		private LoggerInterface $logger,
		private IAppConfig $config,
		?string $userId) {
```

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+use function count;
+use function explode;
+
+class CardDavRateLimitingPlugin extends ServerPlugin {
+
+	private Limiter $limiter;
+	private IUserManager $userManager;
+	private CardDavBackend $cardDavBackend;
+	private IAppConfig $config;
+	private LoggerInterface $logger;
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:47]:
```suggestion
```

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		IUserManager $userManager,
+		CardDavBackend $cardDavBackend,
+		LoggerInterface $logger,
+		IAppConfig $config,
+		?string $userId) {
+		$this->limiter = $limiter;
+		$this->userManager = $userManager;
+		$this->cardDavBackend = $cardDavBackend;
+		$this->config = $config;
+		$this->logger = $logger;
```
Comment (miaulalala) [inline apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:60]:
```suggestion
```

File: apps/dav/tests/unit/CardDAV/Security/CardDavRateLimitingPluginTest.php [hunk 7]
```diff
@@ -0,0 +1,162 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * @copyright 2023 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @author 2023 Christoph Wurst <christoph@winzerhof-wurst.at>
```
Comment (miaulalala) [inline apps/dav/tests/unit/CardDAV/Security/CardDavRateLimitingPluginTest.php:8]: Bonjour

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php (hunk 1 above)
Comment (miaulalala) [reply apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:90]: would be fine if you did :wink:

**Review by kesselb — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

File: apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php (hunk 2 above)
Comment (hamza221) [reply apps/dav/lib/CardDAV/Security/CardDavRateLimitingPlugin.php:75]: yes

**Comment by kesselb** [comment]
My bad, overlooked the existing 29 backport :see_no_evil:

**Pull request merged**